Makefile

```diff
@@ -100,14 +100,34 @@ clean:
 clean-cache:
 	rm -rf ${CACHEDIR}
 
-${CACHEDIR}/.libraries_token.txt: cache
+${CACHEDIR}/.guarded-repo.token: cache
+	tmpf=$(shell mktemp); \
+	chainctl auth octo-sts --identity=guarded-package-repos --scope=chainguard-dev > $${tmpf}; \
+	mv $${tmpf} ${CACHEDIR}/.guarded-repo.token
+
+.PHONY: repo-token
+repo-token: ${CACHEDIR}/.guarded-repo.token
+
+${CACHEDIR}/.libraries.token: cache
 	tmpf=$(shell mktemp); \
 	chainctl auth login --audience libraries.cgr.dev; \
 	chainctl auth token --audience libraries.cgr.dev > $${tmpf}; \
-	mv $${tmpf} ${CACHEDIR}/.libraries_token.txt
+	mv $${tmpf} ${CACHEDIR}/.libraries.token
 
 .PHONY: lib-token
-lib-token: ${CACHEDIR}/.libraries_token.txt
+lib-token: ${CACHEDIR}/.libraries.token
+
+.PHONY: cache-tokens-if-needed/%
+cache-tokens-if-needed/%:
+	auth_needed=$$(yq '.. | select(has("uses")) | select(.uses | test("auth|iamguarded"))' $*); \
+	if [ -n "$$auth_needed" ]; then \
+		$(MAKE) repo-token; \
+		$(MAKE) lib-token; \
```

Comment:
its not clear what the mapping between auth/iamguarded and repo-token/lib-token are.  If there's a 1:1 mapping, can we handle them separately so the relationship is clear? Or do both tokens need to be used by both pipelines?

Also, personally I'd probably just make this a couple of make functions, akin to `authget`.

Reply:
I think the problem we have is with more fine grained filtering, incorporating auth into other pipelines (like the iamguarded pipelines) becomes painful. We have to remember to come back to this and remember to add an exception

I can make this more fine grained and isolated between both use cases. It is something we might want to take another pass at though

If there was a way to "unroll" top level pipelines used in packages, that would make this a little more battle ready

Reply:
Consolidated in a couple functions with a comment on the relationship between iamguarded and auth/guarded-repo